File: leetcode/easy/83-removeDuplicatesFromSortedList.java
class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        ListNode prev = head;
        ListNode curr = head;
        HashSet<Integer> set = new HashSet<>();
        
        while (curr != null) {
            if (set.contains(curr.val)) {
                prev.next = curr.next;
                curr = curr.next;
            } else {
                set.add(curr.val);
                prev = curr;
                curr = curr.next;
            }
        } 
        return head;
    }
}

File: leetcode/easy/83-removeDuplicatesFromSortedList.cpp
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        set<int> seen;
        ListNode *curr = head;
        ListNode *prev = nullptr;

        while (curr != nullptr) {
            cout << curr->val;
            if (seen.find(curr->val) != seen.end()) {
                prev->next = curr->next;
            } else {
                seen.insert(curr->val);
                prev = curr;
            }
            curr = curr->next;
        }
        return head;
    }
};
